### ISSUE TITLE
Journalbeat Stops Reading Journals on Journald Rotation

### ISSUE THREAD

**Issue by joelika** [elastic/beats#9533]
I've been evaluating journalbeat to send data to logstash, and journalbeat stops sending data daily, which I am assuming is tied to when journald is rotating journals. A quick restart of journalbeat gets data sending again.

- Version: 6.5.2
- Operating System: RHEL 7.6

My journald storage mode is set to auto, and I've tried it both with setting up the persistent storage, and without. I've also tried it where I specified the journal location in journalbeat and without.

**Comment by kfalconer** [comment]
Also seeing same issue with 

* Version: 6.5.3
* Operating System: Ubuntu 16.04

**Comment by kristinn** [comment]
I'm able to reproduce using master with https://github.com/elastic/beats/commit/952d92272839248fe8a2cb98dc265ae8c608b674 as the newest commit.

This has been causing us lots of issues since we are processing around 10 million records a day. The "workaround" is to restart journalbeat every 5 minutes.

**Comment by cachedout** [comment]
Possibly related to this: https://github.com/coreos/go-systemd/issues/233

If that's the case, they were waiting on the 1.11 release of golang earlier this year and they may be able to merge their pending fix. I bumped the upstream maintainers just to see.

**Comment by JustinHead** [comment]
We are experiencing this or some other issue that causes journalbeat to stop sending logs on Ubuntu 18.04 which uses systemd v237.

**Comment by mickymiek** [comment]
Same issue here with systemd v238. Restarting journalbeat "solves" the issue.

**Comment by kvch** [comment]
@JustinHead @mickymiek  Do you know what triggers the stopping of shipping logs (e.g. rotate, vacuum)?
I would like to track down whether the issue you are facing is due to a bug in the C API or in Journalbeat.

**Comment by jbguerraz** [comment]
Hello @kvch ,

@mickymiek and I are now investigating the case. We believe it could be related to rotate but have no evidence yet.

During investigations, and probably after deleting journal files, then running ` journalctl --vacuum-time=1seconds`, we started to see tons of `[input]	input/input.go:167	Error while reading event: error while waiting for event: operation not permitted` in logs due to `inotify_init1(O_NONBLOCK|O_CLOEXEC)     = -1 EPERM (Operation not permitted)`

**Comment by kvch** [comment]
This might be due to incorrect seccomp policy. Could you please try to turn seccomp policy to test this theory?
```
seccomp.enabled: false
```

**Comment by jbguerraz** [comment]
Yes. BTW, we're running CoreOS, same case than https://github.com/elastic/beats/pull/10605 so probably covered by your https://github.com/elastic/beats/pull/10593

**Comment by philandstuff** [comment]
I'm experiencing this issue.  I'm running ubuntu 18.04.2 (systemd v237).

Could the issue be that the journalbeat binaries are built on a debian stretch image, and stretch has libsystemd v232?  I wonder if it might be possible to use [libsystemd from backports](https://packages.debian.org/stretch-backports/libsystemd-dev) (which is v239) to get the fixed version?

**Comment by suhailpatel** [comment]
👋, we faced a similar issue on CoreOS Systemd (we reported https://github.com/elastic/beats/pull/10605) and found the result was because the `seccomp` policy was missing `ppoll` (confirmed using strace). 

We've been running with https://github.com/elastic/beats/pull/10593 which fixes the issue 🙌

**Comment by jbguerraz** [comment]
@kvch confirmed, no more issue since we disabled seccomp. Gonna go with the master branch since #10593 got merged. Thx!

**Comment by VTommyV** [comment]
I'm having the same trouble with Journalbeat 6.6.1 on Ubuntu 16.04. Setting "seccomp.enabled: false" has made no difference for me unfortunately. The debug logs show nothing useful, but the journal timestamps indicate that this is happening whenever the journal is rotated.

**Comment by kvch** [comment]
What do you mean by same issue? In this issue, two problem is reported.
What is your systemd version? If it is `v232` or below, journalbeat fails to read entries after rotation, because of a bug in systemd.
If you are seeing operation not permitted errors, that is due to a bug in journalbeat, which is going to be fixed in the next release. As a workaround you can set `seccomp.enabled` to `false`.

**Comment by VTommyV** [comment]
@kvch the issue being that Journalbeat stops shipping data. Ubuntu 16.04 uses systemd 229 and a future backport of that fix will be unlikely in an LTS release. Will [#10593](https://github.com/elastic/beats/pull/10593) help me with this?

**Comment by kvch** [comment]
Unfortunately, not. The systemd API call we are using right now has the bug. But in a future release, we would like to refactor the code to remove this dependency.

**Comment by hamelg** [comment]
The same here for centos/RHEL 7.6. Is there a related open issue about that ?

**Comment by kvch** [comment]
I have opened an issue to track the two problems in different issues: https://github.com/elastic/beats/issues/11110

**Comment by hamelg** [comment]
> Unfortunately, not. The systemd API call we are using right now has the bug. But in a future release, we would like to refactor the code to remove this dependency.

Sorry, but my question was about the bug in API systemd <=v232 using by journalbeat. 
Centos/RHEL 7.6 runs systemd v219. Is there a chance to see a fix in a near future ?

**Comment by kvch** [comment]
This is the issue to track that problem. It is definitely something we want to fix in an upcoming release.

**Comment by hamelg** [comment]
So, its status shouldn't be closed.

**Comment by kvch** [comment]
Reopened.

**Comment by liqw33d** [comment]
It looks like that: https://github.com/coreos/go-systemd/pull/279 is fixing this problem.
I re-compiled journald with this patch applied and now there are no problems when systemd journal is rotated. Test is running ~2days and all messages were processed.
No other regressions were found during this testing.
Tested system is Ubuntu 16.04 (xenial) and Go is 1.11.

**Comment by kvch** [comment]
How did you apply the patch? We use `github.com/coreos/go-systemd/sdjournal` not `github.com/coreos/go-systemd/journal`. Did you copy the latter package also?
In the meantime I am looking into it.

**Comment by kvch** [comment]
Also, what's your systemd version?

**Comment by kvaps** [comment]
@kvch I was just replaced whole `go-systemd` directory with newer version.
You can try one of my docker image:
```
kvaps/journalbeat:7.0.0-fix9533
kvaps/journalbeat:8.0.0-fix9533
```

**Comment by kvaps** [comment]
Agree. This fix is working fine for me!

> what's your systemd version?
```
- systemd 229 (xenial)
- systemd 232 (stretch)
- systemd 237 (bionic)
- systemd 241 (stretch)

**Comment by liqw33d** [comment]
> Also, what's your systemd version?

systemd version is:
`229-4ubuntu21.17` (xenial)
I applied mentioned diff to `go-systemd` (master) and compiled `journalbeat` (master) with it's dependencies.
I don't have exact steps to reproduce it now. 
```
  "system_info": {
    "build": {
      "commit": "005bb608b010fa777db527738ad27b5f40086986",
      "libbeat": "8.0.0",
      "time": "2019-03-16T22:51:04.000Z",
      "version": "8.0.0"
    }
  }
```

**Comment by vquiering** [comment]
When will this be available through the official elastic repo?

**Comment by kvch** [comment]
@liqw33d  @kvaps Thank you for verifying the fix!

@vquiering Once the fix is released in a new version of `github.com/coreos/go-systemd`, I am updating the vendored folder.

**Comment by hamelg** [comment]
> I re-compiled journald with this patch applied and now there are no problems when systemd journal is rotated. Test is running ~2days and all messages were processed.

Does it mean the fix consists to recompile journald ? The patch you mentioned is only related to github.com/coreos/go-systemd/journal (for writing to systemd's logging service, journald), journalbeat uses github.com/coreos/go-systemd/sdjournal (for reading from journald by wrapping its C API). Please, can you clarify ?

**Comment by kvaps** [comment]
@hamelg yes it is, you just need to recompile journalbeat with updated go-systemd

**Comment by stanxing** [comment]
> yes it is, you just need to recompile journalbeat with updated go-systemd
I feel a little confused. Why it can work when updating the package that has not been depended by journalbeat?

**Comment by kvch** [comment]
This is what I am investigating. My hunch is that the `init` function of that package connects to the FD of systemd journal, thus it lets Journalbeat read when and what changes are being made to the journal it is reading from.

**Comment by kvch** [comment]
I did more investigation on the matter. It turns out that the problem was fixed in Journalbeat not in its dependency `github.com/coreos/go-systemd/sdjournal`. The upcoming 6.7 and 7.0.0-rc release works with `systemd v232`.

Do you mind verifying my findings once the release is available?

**Comment by stanxing** [comment]
> The upcoming 6.7 and 7.0.0-rc release works with `systemd v232`.

Take the liberty to ask, when is the 6.7 and 7.0.0-rc release date ?

**Comment by praseodym** [comment]
6.7 has just been released.

**Comment by stanxing** [comment]
the release of 6.7 works fine for me  on ubuntu 16.04  with systemd version 229。But I found some erros as f:

**Comment by hamelg** [comment]
I confirm 6.7 has fixed the issue on centos 7.6 with systemd version 219.

**Comment by hamelg** [comment]
> But I found some erros as follows:
> 
> ```
> 2019-03-27T17:52:36.001+0800    ERROR   [input] input/input.go:167      Error while reading event: failed to get realtime timestamp: 99  {"id": "a6a84c06-f11d-4f37-9d8f-3251d731d235"}
> ```

The same here, I see this error each time the journal is rotating.

**Comment by kvch** [comment]
@hamelg  Do you mind opening a separate issue for that? Thanks for the confirmation.

**Comment by kvch** [comment]
Fixed in 6.7 and on master.
